`Clear-Site-Data: "cache"` HTTP header should clear disk cache more efficiently

cdumez opened the pull request:
#### 28678cd68b668ca4d14eec8a3a1775374f304804
<pre>
`Clear-Site-Data: &quot;cache&quot;` HTTP header should clear disk cache more efficiently
<a href="https://bugs.webkit.org/show_bug.cgi?id=251768">https://bugs.webkit.org/show_bug.cgi?id=251768</a>

Reviewed by Antti Koivisto.

`Clear-Site-Data: &quot;cache&quot;` HTTP header should clear disk cache more efficiently.
We will only delete cache entries for a given partition so there is no point in
traversing ALL storage entries. Instead, only traverse entries for the given
partition, which is a lot faster.

This patch adds a new NetworkCache::traverse() function overload which takes in
the cache partition as parameter and only calls the traverseHandler for entries
belonging to this partition. This is cheap and easy to do for our disk cache
implementation since our cache entries as grouped by partition via subfolders.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::traverse):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::Key::partitionToPartitionHashAsString):
(WebKit::NetworkCache::Key::computePartitionHash const):
(WebKit::NetworkCache::Key::partitionToPartitionHash):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::traverseInternal):
(WebKit::NetworkCache::Storage::traverse):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:

Canonical link: <a href="https://commits.webkit.org/259902@main">https://commits.webkit.org/259902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4196e8fe9988564ee822422e7af0f97ac2727436

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106346 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15400 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115539 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6617 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115226 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112110 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12815 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95799 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40375 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94690 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27442 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8636 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28794 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9148 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5364 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48340 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10710 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3686 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -969,7 +969,7 @@ void Storage::store(const Record& record, MappedBodyHandler&& mappedBodyHandler,
 }
 
-void Storage::traverse(const String& type, OptionSet<TraverseFlag> flags, TraverseHandler&& traverseHandler)
+void Storage::traverseInternal(const String& rootPath, const String& type, OptionSet<TraverseFlag> flags, TraverseHandler&& traverseHandler)
```
I don't like calling things FooInternal. Can this say something about what it actually does like `traverseWithinRootPath`?

cdumez replied:
Sounds good.

cdumez committed:
`Clear-Site-Data: "cache"` HTTP header should clear disk cache more efficiently
https://bugs.webkit.org/show_bug.cgi?id=251768

Reviewed by Antti Koivisto.

`Clear-Site-Data: "cache"` HTTP header should clear disk cache more efficiently.
We will only delete cache entries for a given partition so there is no point in
traversing ALL storage entries. Instead, only traverse entries for the given
partition, which is a lot faster.

This patch adds a new NetworkCache::traverse() function overload which takes in
the cache partition as parameter and only calls the traverseHandler for entries
belonging to this partition. This is cheap and easy to do for our disk cache
implementation since our cache entries as grouped by partition via subfolders.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::traverse):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::Key::partitionToPartitionHashAsString):
(WebKit::NetworkCache::Key::computePartitionHash const):
(WebKit::NetworkCache::Key::partitionToPartitionHash):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::traverseInternal):
(WebKit::NetworkCache::Storage::traverse):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:

Canonical link: https://commits.webkit.org/259902@main
